File: Lintcode 015 - Permutations.java
Lintcode 015 - Permutations.java

class Solution {
    /**
     * @param nums: A list of integers.
     * @return: A list of permutations.
     */
    public ArrayList<ArrayList<Integer>> permute(ArrayList<Integer> nums) {
        // write your code here
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if(nums == null || nums.size() == 0) {
            return result;
        }
        
        ArrayList<Integer> list = new ArrayList<Integer>();
        dfs(result, list, nums);
        return result;
    }
    
    public void dfs(ArrayList<ArrayList<Integer>> result, ArrayList<Integer> list, ArrayList<Integer> nums) {
        
        if(list.size() == nums.size()) {
            result.add(new ArrayList<Integer>(list));
            return;
        }
        
        for(int i = 0; i < nums.size(); i++) {
            if(!list.contains(nums.get(i))) {
                list.add(nums.get(i));
                dfs(result, list, nums);
                list.remove(list.get(list.size()-1));
            }
        }
    }
}

File: Lintcode 015 - Permutations.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace CSharp
{
    class Program{
        static void Main()
        {
            int[] nums = new int[4]{1,2,3,4};
            List<List<int>> result = new List<List<int>>{};   
            List<int> list = new List<int>{};

            Permutations(nums, result, list);

            foreach(var clist in result)
            {
                foreach(var num in clist)
                {
                    Console.Write(num + ", ");
                }
                Console.WriteLine(" ");
            }
        }

        static void Permutations(int[] nums, List<List<int>> result, List<int> list)
        {
            if(list.Count == nums.Length)
            {
                result.Add(new List<int>(list));
                return;
            }

            for(int i = 0; i< nums.Length; i++)
            {
                if(!list.Contains(nums[i]))
                {
                    list.Add(nums[i]);
                    Permutations(nums, result, list);
                    list.Remove(list[list.Count - 1]);
                }
            }
        }
    }
}
